ch5-catdir/directory.cpp

```diff
@@ -48,14 +49,60 @@ std::optional<Directory::Entry> Directory::next()
                             "failed to read entry from " + m_path);
 }
 
+namespace
+{
+
+/*
+ * readlink - reads the content of the symlink 'path', using initial buffer of
+ *            size 'size'
+ *
+ * We can't rely on the size returned by lstat, because the link can increase
+ * between tha calls, and because the kernel doesn't give a size for the
+ * symlinks on pseudo-filesystems such as /proc, so we need to iterate until the
+ * buffer fits the link content.
+ *
+ * The code main layout is based on the GNU coreutils v5.0 xreadlink function.
+ */
+
+std::optional<std::string> readlink(const std::string& path, off_t size)
+{
+    if (size == 0) // the OS doesn't give us the size, so we need to guess
+    {
+        size = 80;
+    }
+    auto res = std::string();
+    while (true)
+    {
+        res.resize(size + 1); // one additional character for the terminating
+                              // null
+        if (auto readCount = ::readlink(path.c_str(), res.data(), size + 1);
+            readCount < 0)
+        {
+            auto savedErrno = errno;
+            Controller::error() << "failed to read link " << path << ": "
+                                << std::generic_category().message(savedErrno)
+                                << '\n';
+            return std::nullopt;
+        }
+        else if (readCount <= size)
+        {
+            res.resize(readCount + 1);
+            return res;
+        }
+        size *= 2;
+    }
+}
+
+}
+
 Directory::Entry Directory::entryFor(std::string name)
 {
-    auto res = Entry{ std::move(name), std::nullopt };
+    auto res = Entry{ std::move(name), std::nullopt, std::nullopt };
```

Comment:
Do you really need to pass `std::nullopt` explicitly?

Reply:
gcc and clang warn about this.